### ISSUE TITLE
Websockets jsr356 willDecode not invoked during decoding

### ISSUE THREAD

**Issue by melowe** [eclipse/jetty.project#3389]
If you have several decoders registered on a ClientEndpoint the decoders are exerted regardless of the willDecode function's outcome. As far as I can see willDecode function isn't executed at all.

**Comment by joakime** [comment]
This would be surprising, and if replicated, a bug that wasn't caught in our test cases.

What type of decoder are you using? 
What's the signature of your decoder?
And how are you declaring the decoders for your ClientEndpoint? (configurator? annotations? other?)

**Comment by joakime** [comment]
Also, what version of Jetty?

**Comment by joakime** [comment]
Opened PR #3419

**Comment by joakime** [comment]
@melowe we are still waiting for your use case, how you can possibly have multiple Decoders on a `@ClientEndpoint`.
The use case for willDecode is to find the correct decoder to use for your `@OnMessage` declaration.
Jetty figures that out at Endpoint initialization, there is only ever one Decoder assigned to a single `@OnMessage` (that is chosen from your declared list of Decoders).

**Comment by melowe** [comment]
I'll add a test case when I get some time. But basically I have an annotated message handler that accepts object, and then I register multiple decoders where the willDecode is used to decode or not depending on the payload. This approach worked with Tyrus. when I stepped through the code I was the willDecode was never being invoked and therefore the incorrect decoder was being called. 

I assume the PR addresses this issue?

**Comment by joakime** [comment]
The PR #3419 does, and it's been merged into `jetty-9.4.x` (the implementation in `jetty-10.0.x` already had willDecode support).
Incidentally, you are officially the first person that I've come across that uses Decoder lists and cares about `willDecode`.

What's interesting is that the [TCK for javax.websocket](https://github.com/eclipse-ee4j/jakartaee-tck/tree/master/src/com/sun/ts/tests/websocket) has a few examples of Decoder lists, but few of them are valid as a list.

They do things like have `decoders = { StringBeanBinaryStreamDecoder.class, StringBeanTextStreamDecoder.class, StringBeanBinaryDecoder.class, StringBeanTextDecoder.class }`
That's a nonsensical decoder list, as `Decoder.BinaryStream` and `Decoder.TextStream` have no `willDecode`, meaning that the last 2 entries will never be used.  So why include them??  This should actually fall under the part of the spec where we fail to add an endpoint (and cause a failure to deploy) if there is something amiss with its configuration.
Our support for Decoder lists is also not right, so I opened #3428 
The Decoder list support in the TCK is also incomplete, as the search order within the various lists isn't defined by the spec or the TCK (looking at the various implementations out there, we all do it differently, yet we all pass the TCK), so I opened eclipse-ee4j/websocket-api#289

Honestly, the whole willDecode is odd, and Decoder list is odd, and so wasteful of resource (your GC churn must be huge).
But it exists in the API, so we should make a better effort at supporting it.
It's probably better if you just use a single Decoder that can return a variety of objects that extend from the one you have declared in your `@OnMessage` annotated method handler, or `MessageHandler`, you'll have less object churn and a more consistent experience across all of the various javax.websocket implementations.  
This rarely used feature of javax.websocket might be more consistent in the future once we can work out a better set of TCK tests for it with the rest of the implementations.
